pilot/pkg/model/xds_cache.go

```diff
@@ -222,36 +210,39 @@ func (l *lruCache) assertUnchanged(key string, existing *discovery.Resource, rep
 	}
 }
 
-func (l *lruCache) Add(entry XdsCacheEntry, token CacheToken, value *discovery.Resource) {
-	if !entry.Cacheable() {
+func (l *lruCache) Add(entry XdsCacheEntry, pushReq *PushRequest, value *discovery.Resource) {
+	if !entry.Cacheable() || pushReq == nil {
 		return
 	}
+	// Start unset
+	if pushReq.Start.Equal(time.Time{}) {
+		panic("token cannot be empty!")
+	}
+	// TODO: It will not overflow until year 2262
+	token := CacheToken(pushReq.Start.UnixNano())
 	l.mu.Lock()
 	defer l.mu.Unlock()
 	k := entry.Key()
 	cur, f := l.store.Get(k)
-	toWrite := cacheValue{value: value}
 	if f {
-		if token != cur.(cacheValue).token {
+		// This is the stale or same resource
+		if token <= cur.(cacheValue).token || token <= l.token {
 			// entry may be stale, we need to drop it. This can happen when the cache is invalidated
 			// after we call Get.
 			return
 		}
-		// Otherwise, make sure we write the current token again. We don't change the key on writes; the
-		// same token will be used for a value until its invalidated
-		toWrite.token = cur.(cacheValue).token
-	} else {
-		// This is our first time seeing this; this means it was invalidated recently and this is our
-		// first write, or we forgot to call Get before.
-		return
-	}
-	if l.enableAssertions {
-		if toWrite.token == 0 {
-			panic("token cannot be empty. was Get() called before Add()?")
+		if l.enableAssertions {
+			l.assertUnchanged(k, cur.(cacheValue).value, value)
```

Comment:
I think this no longer does anything since L229 should always cause it to return

Reply:
After fixing this to assert properly (I think - maybe my change was bad) by changing `<=` to `<`, it failed a lot

Reply:
Then we need some other way to trigger the assertUnchanged. Right now it will never be triggered, so we have no test coverage of the change here?

Reply:
looking into this

Reply:
This is caused by the bugs from other pr, which are fixed now